Continued Work on Advanced Indexing

reminisce opened the pull request:
## Description ##
This is a continued work of https://github.com/apache/incubator-mxnet/pull/7910 addressing https://github.com/apache/incubator-mxnet/issues/8084.
Implementation based upon Numpy's [indexing documentation](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html#advanced-indexing).

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support NDArray `__setitem__` and `__getitem__` using advanced indices **excluding** Numpy Ellipsis(...), newaxis, and [boolean array indexing](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html#boolean-array-indexing). When index is a single list, only a list of integers is supported. Nested lists as indices is supported when index is a tuple of basic and advanced indices. That is, `a[[1, 2]]` is supported, `a[[[1], [2]]]` is NOT supported, but `a[1, [[1], [2]], [1, 2], 2:5]` is supported.
- [x] Support NDArray indexing using slices with arbitrary value of `step`.
- [x] Re-implemented operators `slice`, `_slice_assign`, and `_slice_assign_scalar` using `Kernel::Launch` and support `step != 1`.
- [x] Fixed this issue: https://github.com/apache/incubator-mxnet/issues/8164

@piiswrong @eric-haibin-lin @szha

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+                if key.step == 0:
+                    raise ValueError("slice step cannot be zero")
+                start, stop, step = _get_index_range(key.start, key.stop, self.shape[0], key.step)
+                indices = [i for i in range(start, stop, step)]
```
use nd.arange

reminisce replied:
Done.

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+                return self.take(array(indices, ctx=self.context, dtype=np.int32))
+            elif key.start is not None or key.stop is not None:
                 return self._slice(key.start, key.stop)
-            return self
```
There isn't a else block at the end. Is this intentional?

piiswrong replied:
never mind .

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+        # if key is any type of NDArray, list, or np.ndarray make a tuple from it
+        # Note: Only support integer list as index if key is a list.
+        if isinstance(key, (NDArray, np.ndarray)):
+            key = (key,)
```
Go to advanced indexing directly?

reminisce replied:
I refactored this part as you suggested to make the logic more clear.

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+            slices and integers."""
+            return not isinstance(index, py_slice)
+
+        if not isinstance(key, tuple):
```
Replace with simple assert since this is never reached.

reminisce replied:
Done.

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
+                idx_i = arange(start, stop, step, ctx=self.context, dtype=dtype)
+                advanced_indices.append(i)
+            elif isinstance(idx_i, NDArray):
+                dtype = idx_i.dtype
```
different indices could have inconsistent type.

reminisce replied:
I forced all dtypes to be 'int32' and added a unit test case.

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +538,260 @@ def __getitem__(self, key):
+            oshape.append(1)
+        return op.slice(self, begin, end).reshape(oshape)
+
+    def _basic_indexing_step_not_one(self, key):
```
This can be combined with _basic_indexing_step_equal_one by adding step to op.slice

reminisce replied:
I will submit a separate PR implementing slice op with step to cover this change.

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -1703,6 +1912,109 @@ def tostype(self, stype):
+    elif stop < 0:
+        stop += length
+        if stop < 0:
+            raise IndexError('Slicing stop %d exceeds limit of %d' % (stop-length, length))
```
does numpy raise error for this?

reminisce replied:
Yes. See below:
```python
>>> import numpy as np
>>> a = np.array([[1, 2], [3, 4]])
>>> a[-3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index -3 is out of bounds for axis 0 with size 2
```

piiswrong commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -1703,6 +1912,109 @@ def tostype(self, stype):
+    return start, stop, step
+
+
+def _get_broadcast_shape(shape1, shape2):
```
Does numpy have a util for this?

reminisce replied:
I hope so, but I tried searching online and couldn't find one in numpy taking shapes as input arguments.

jmargeta replied:
If that is of any use, there is a private function in stride_tricks:
https://github.com/numpy/numpy/blob/77f9540f277de3e727d696eb5987a0505ed8cdf9/numpy/lib/stride_tricks.py#L176

reminisce replied:
@jmargeta Thanks for the information. It looks like that the function in Numpy literally broadcasts ndarrays to get the final shape, which is not efficient and a overkill in this case. Also when the current MXNet NDArray is allocated on GPU, that function does not work. What we need is only the final shape as an input to the MXNet broadcast_to operator which is more efficient and works on both CPU and GPU.

JeanKossaifi commented:
Awesome changes!
Is this ready to be merged?

reminisce commented:
@JeanKossaifi Thanks for the comment. This only implements `__getitem__` of `NDArray`. We will add support for `__setitem__` later and add it to the PR.

piiswrong commented on src/operator/tensor/indexing_op.h:
```diff
@@ -880,6 +880,19 @@ void ScatterNDForward(const nnvm::NodeAttrs& attrs,
+ * DO NOT use call this function unless you have to.
+ */
+template<typename xpu>
+void _ScatterSetNDForward(const nnvm::NodeAttrs& attrs,
```
no need for _ here

piiswrong commented on src/operator/tensor/indexing_op.cc:
```diff
@@ -355,6 +355,36 @@ Examples::
+.set_attr<nnvm::FInferShape>("FInferShape", ScatterNDShape)
+.set_attr<nnvm::FInferType>("FInferType", ScatterNDType)
+.set_attr<FCompute>("FCompute<cpu>", _ScatterSetNDForward<cpu>)
+.set_attr<nnvm::FGradient>("FGradient", MakeZeroGradNodes)
```
Shouldn't have gradient

szha commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -507,71 +509,357 @@ def __getitem__(self, key):
+                return self
+
+        # Start dealing with more complicated indexing cases than only slicing at axis=0.
+        # if key is any type of NDArray, list, or np.ndarray make a tuple from it
```
can the key be other iterable such as a generator expression?

reminisce replied:
Are you referring to something like this: `a[i for i in range(3)]`?
I tried it in numpy, it doesn't work.
```python
>>> import numpy as np
>>> a = np.arange(16).reshape(4, 4)
>>> a
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])
>>> a[x for x in range(3)]
  File "<stdin>", line 1
    a[x for x in range(3)]
          ^
SyntaxError: invalid syntax
```

dai-dao commented:
Hi, I'm aware that this issue is currently being developed, but is there any temporary workaround I can use in the mean time? What I'm trying to do is this:

Thanks

```
a = nd.array([[1,2,3], [4,5,6]])
a[:, [0, 1]]
```

dai-dao commented:
Hi, I'm aware that this issue is still being implemented, but I want to ask for a viable workaround if there's any. Here's what I want to do:

```
>>> from mxnet import nd
>>> a = nd.array([[1,2], [3, 4]])
>>> a

[[ 1.  2.]
 [ 3.  4.]]
<NDArray 2x2 @cpu(0)>
>>> 
>>> a[[0, 1], [0, 1]]

[ 1.  4.]
```

Now I want to assign values to these indexes:

```
>>> a[[0, 1], [0, 1]] = [5, 6]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/daidao/anaconda3/envs/rl/lib/python3.6/site-packages/mxnet/ndarray/ndarray.py", line 440, in __setitem__
    str(slice_i), str(type(slice_i))))
ValueError: NDArray does not support slicing with key [0, 1] of type <class 'list'>.
```

and I got an error. Is there a way to work around this for the mean time? Thank you

reminisce commented:
@dai-dao You can use [`gather_nd`](https://mxnet.incubator.apache.org/versions/master/api/python/ndarray/ndarray.html#mxnet.ndarray.gather_nd) op.

```python>>> import mxnet as mx
>>> a = mx.nd.array([[1,2], [3, 4]])
>>> index = mx.nd.array([[0, 1], [0, 1]], dtype='int32')
>>> mx.nd.gather_nd(a, index)

[ 1.  4.]
<NDArray 2 @cpu(0)>
```

dai-dao commented:
Hi,

That is for getting the values, which works fine using the index, however setting values using the index is what causing the error. I don't understand how gather_nd would solve it. Could you elaborate?

reminisce commented:
@dai-dao There is an op called `scatter_nd` which can do the trick. However, this op will reset all the unindexed elements in the output to zero. So this might not be a desirable solution to you.
```python
>>> import mxnet as mx
>>> a = mx.nd.array([[1,2], [3, 4]])
>>> index = mx.nd.array([[0, 1], [0, 1]], dtype='int32')
>>> data = mx.nd.array([5, 6])
>>> mx.nd.scatter_nd(data, index, shape=(2, 2), out=a)
>>> a
[[ 5.  0.]
 [ 0.  6.]]
<NDArray 2x2 @cpu(0)>
```
If you really want to use advanced indexing, you can checkout the branch of this PR and start using advanced indexing to write code. The work is mostly done. We just need to add more tests and make the logic dispatching cleaner.

reminisce commented:
@piiswrong @eric-haibin-lin @szha I added support for `__setitem__` and simplified basic and advanced indexing dispatching logic. Please review again. Thanks.

meanmee commented:
could this feature be used in version 0.12？

reminisce commented:
@meanmee This is a feature for 1.0 which is coming soon. Thanks for your interest.

JeanKossaifi commented:
Thanks for the great work @reminisce !

cjolivier01 commented:
@reminisce Does this intentionally remove the _internal_full() operator and return to old-style create-with-empty + assign?

cjolivier01 commented on python/mxnet/ndarray/ndarray.py:
```diff
@@ -1858,10 +2174,8 @@ def full(shape, val, ctx=None, dtype=mx_real_t, out=None):
-        ctx = Context.default_ctx
-    dtype = mx_real_t if dtype is None else dtype
-    out = _internal._full(shape=shape, ctx=ctx, dtype=dtype, value=val, out=out)
+    out = empty(shape, ctx, dtype) if out is None else out
```
Why revert?

reminisce replied:
Don't remember exactly. @piiswrong once asked me to change something back, not sure if it's this one. But one benefit of changing it is allowing `val` to be any data types that could be assigned to an `NDArray`. Behavior-wise, it's same as before when val is a scalar. Any problem with this change?

cjolivier01 replied:
@piiswrong do you recall?

reminisce replied:
@cjolivier01 here is the link to the conversation.
https://github.com/apache/incubator-mxnet/pull/8340/files/3f88ddf475a5f9920b7e11b606579cc8fc509edb#diff-151ed3672c6bd71b358e13771357696dR1860

cjolivier01 replied:
ok thanks for digging that up!

reminisce commented:
@cjolivier01 For assigning a value to a whole array, it still calls `_internal._full`. Are there any other places you expect to call `_full`?

cjolivier01 commented:
just there
